### PR TITLE
Erpgrid

### PR DESCRIPTION
**Pull request by vladdez**

### CODE REVIEW

**Comment by vladdez** [comment]
Could you please look what's happened here - here was a huge merge conflict, I resolved it as i could, but not sure that it was correct (I didn't know about bunch of changes)

**Comment by vladdez** [comment]
everything until "small test" was mine

File: src/plot_erpgrid.jl [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+# no figure?
+plot_erpgrid(plotData::Matrix{<:Real}, pos; kwargs...) =
+    plot_erpgrid!(Figure(), plotData, pos; kwargs...)
+
+function plot_erpgrid!(
+    f::Union{GridPosition,Figure},
+    plotData::Matrix{<:Real},
+    pos;
+    chanNum = 30,
```
Comment (behinger) [inline src/plot_erpgrid.jl:28]: why do we need chanNum? we can always do size(plotData,1)

File: src/plot_erpgrid.jl [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+The input `f`
+"""
+
+# no figure?
+plot_erpgrid(plotData::Matrix{<:Real}, pos; kwargs...) =
+    plot_erpgrid!(Figure(), plotData, pos; kwargs...)
+
+function plot_erpgrid!(
+    f::Union{GridPosition,Figure},
+    plotData::Matrix{<:Real},
```
Comment (behinger) [inline src/plot_erpgrid.jl:26]: some functions work right now on dataframes, some on Matrix.
We should unify this. Unfold.jl by default returns DataFrames (via `coeftable`) which are basically linearized Matrices (e.g. plotData[:]). Not 100% sure how to best handle it, but it is always annoying to need to convert to the other format :)

File: src/plot_erpgrid.jl [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+## Return Value:
+The input `f`
+"""
+
+# no figure?
+plot_erpgrid(plotData::Matrix{<:Real}, pos; kwargs...) =
+    plot_erpgrid!(Figure(), plotData, pos; kwargs...)
+
+function plot_erpgrid!(
+    f::Union{GridPosition,Figure},
```
Comment (behinger) [inline src/plot_erpgrid.jl:25]: this is outside of this PR, but a new one:
Here the method signatures are defined, and we are not defining them correctly ^^ 
https://docs.makie.org/stable/explanations/plot_method_signatures/

File: src/plot_erpgrid.jl (hunk 1 above)
Comment (behinger) [reply src/plot_erpgrid.jl:28]: if you want to restrict the channels to be plotted, you'd rather expect a range/vector of channels to be used, else you can only restrict up to chanNum. I would remove this functionality and let the user subset channels / positions themselves

File: src/plot_erpgrid.jl [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    chanNum = 30,
+    drawLabels = false,
+    times = 1:size(plotData, 2),
+    kwargs...,
+)
+    plotData = plotData[1:chanNum, :]
+    pos = hcat([[p[1], p[2]] for p in pos]...)
+
+    pos = pos[:, 1:chanNum]
+    minmaxrange = (maximum(pos, dims = 2) - minimum(pos, dims = 2))
```
Comment (behinger) [inline src/plot_erpgrid.jl:37]: we should move this standardization to an extra function. We need it in other places as well

File: src/plot_erpgrid.jl [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    rel_zeropoint = argmin(abs.(times)) ./ length(times)
+
+    for (ix, p) in enumerate(eachcol(pos))
+        x = p[1] #- 0.1
+        y = p[2] #- 0.1
+        # todo: 0.1 should go into plot config
+        ax = Axis(
+            f[1, 1],
+            width = Relative(0.2),
```
Comment (behinger) [inline src/plot_erpgrid.jl:51]: the 0.2 should be a configurable quantitiy in the "visual" category

File: src/plot_erpgrid.jl [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            height = Relative(0.2),
+            halign = x,
+            valign = y,
+        )# title = raw_ch_names[1:30])
+        if drawLabels
+            text!(
+                ax,
+                rel_zeropoint + 0.1,
```
Comment (behinger) [inline src/plot_erpgrid.jl:59]: the distance of the label should be configurable imho - in general all of these settings should be configurable, we can achieve this by either using a PlotConfig category, or by an argument::NamedTuple and then merging our defaults wiht the provided defaults

File: src/plot_erpgrid.jl [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                align = (:left, :top),
+                space = :relative,
+            )
+        end
+        # todo: add label if not nothing
+
+        push!(axlist, ax)
+    end
+    # todo: make optional + be able to specify the linewidth + color
+    hlines!.(axlist, Ref([0.0]), color = :gray, linewidth = 0.5)
```
Comment (behinger) [inline src/plot_erpgrid.jl:73]: same as above, but you already have a todo there :)

File: src/plot_erpgrid.jl [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    times = isnothing(times) ? (1:size(plotData, 2)) : times
+
+    # todo: add customizable kwargs
+    h = lines!.(axlist, Ref(times), eachrow(plotData))
+
+    linkaxes!(axlist...)
+    hidedecorations!.(axlist)
+    hidespines!.(axlist)
+
+    f
```
Comment (behinger) [inline src/plot_erpgrid.jl:85]: maybe return not only f, but also the axlist?

**Review by behinger — COMMENTED** [review]
Nice start! I have some comments for you :)

File: src/plot_erpgrid.jl (hunk 3 above)
Comment (vladdez) [reply src/plot_erpgrid.jl:25]: what do you mean exactly?

File: src/plot_erpgrid.jl (hunk 8 above)
Comment (vladdez) [reply src/plot_erpgrid.jl:85]: why?

Comment (behinger) [reply src/plot_erpgrid.jl:85]: because we do the same in eeg_topolotseries - or whatever the function is called

Comment (behinger) [reply src/plot_erpgrid.jl:85]: maybe we should look at all input + output-types and check what makes sense. a FigureAxisPlot object is on the table as well, but related to the mutating/non-mutating discussion

**Comment by vladdez** [comment]
okay let's include this changes for now
there a lot do with every function, the number of bugs is just insane

**Pull request merged**